omnixai/explainers/ranking/agnostic/validity.py

```diff
@@ -0,0 +1,215 @@
+        self.median_features = self._compute_stats(
+            self.training_data, self.features, "median"
+        )
+
+        if preprocessing_function:
+            self.preprocessing_fn = preprocessing_function
+        else:
+            self.preprocessing_fn = lambda x: x.to_pd()
+        self.predict_fn = predict_function
+
+    @staticmethod
+    def _compute_stats(data: pd.DataFrame, features: List, mask_type):
+        if mask_type == "mean":
+            stats = data.mean()
+        if mask_type == "median":
+            stats = data.median()
+
+        for f in features:
+            if f not in stats:
+                most_common_value = data[f].value_counts().idxmax()
+                stats[f] = most_common_value
+
+        return stats
+
+    def _compute_mask(self, mask, idx):
+        if mask == "mean":
+            return self.mean_features[self.features[idx]]
+        elif mask == "median":
+            return self.median_features[self.features[idx]]
+        else:
+            return 0.0
+
+    @staticmethod
+    def _create_copy(sample):
+        if type(sample) == np.ndarray or type(sample) == pd.DataFrame:
+            x = sample.copy()
+        else:
+            x = sample.clone()
+        return x
+
+    def compute_mask(self, x, mask, idx):
+        if type(x) == np.ndarray:
+            x[:, :, idx] = self._compute_mask(mask, idx)
+        elif type(x) == pd.DataFrame:
+            x[self.features[idx]] = self._compute_mask(mask, idx)
+        else:
+            raise Exception("input must be either numpy array or pandas DataFrame")
+        return x
+
+    @staticmethod
+    def compute_pairs(n_docs):
+        positions = list(range(1, n_docs + 1))
+        combs = list(itertools.combinations(positions, r=2)) + list(
+            itertools.combinations(positions[::-1], r=2)
+        )
+        pairs = np.array(list(filter(lambda x: x[0] < x[1], combs)))
+        return pairs
+
+    @staticmethod
+    def compute_rank(scores):
+        return (-scores).flatten().argsort().argsort() + 1
+
+    @staticmethod
+    def calculate_propensity(scores, ranks, i, j):
+        return (scores[i] - scores[j]) * abs(ranks[i] - ranks[j])
+
+    def compute_validity(self, pi, minimal_features, mask, sample, categorical_cols):
+        x = self._create_copy(sample)
+        for i in range(0, len(self.features)):
+            if i not in minimal_features:
+                x = self.compute_mask(x, mask, i)
+        scores = self.predict_fn(
+            Tabular(x, categorical_columns=categorical_cols)
+        ).flatten()
+        print(scores)
+        ranks = self.compute_rank(scores)
+        print(ranks)
+        return {
+            'Tau': scipy.stats.kendalltau(ranks, pi),
+            'Weighted_Tau': scipy.stats.weightedtau(ranks, pi),
+            'Top_K_Ranking': ranks,
+            'Ranks': pi
+        }
+
+    @staticmethod
+    def compute_n_docs(tabular_data: Tabular):
+        return tabular_data.to_pd().shape[0]
```

Comment:
Tabular instance has shape function. Replace this line with tabular_data.shape[0].

Reply:
Done